aptos-move/framework/aptos-framework/sources/staking_contract.move

```diff
@@ -0,0 +1,1132 @@
+            &mut store.add_distribution_events,
+            &mut store.request_commission_events,
+        );
+    }
+
+    fun request_commission_internal(
+        operator: address,
+        staking_contract: &mut StakingContract,
+        add_distribution_events: &mut EventHandle<AddDistributionEvent>,
+        request_commission_events: &mut EventHandle<RequestCommissionEvent>,
+    ): u64 {
+        // Unlock just the commission portion from the stake pool.
+        let (total_active_stake, accumulated_rewards, commission_amount) =
+            get_staking_contract_amounts_internal(staking_contract);
+        staking_contract.principal = total_active_stake - commission_amount;
+
+        // Short-circuit if there's no commission to pay.
+        if (commission_amount == 0) {
+            return 0
+        };
+
+        // Add a distribution for the operator.
+        add_distribution(operator, staking_contract, operator, commission_amount, add_distribution_events);
+
+        // Request to unlock the commission from the stake pool.
+        // This won't become fully unlocked until the stake pool's lockup expires.
+        stake::unlock_with_cap(commission_amount, &staking_contract.owner_cap);
+
+        let pool_address = staking_contract.pool_address;
+        emit_event(
+            request_commission_events,
+            RequestCommissionEvent { operator, pool_address, accumulated_rewards, commission_amount },
+        );
+
+        commission_amount
+    }
+
+    /// Staker can call this to request withdrawal of part or all of their staking_contract.
+    /// This also triggers paying commission to the operator for accounting simplicity.
+    public entry fun unlock_stake(staker: &signer, operator: address, amount: u64) acquires Store {
+        // Short-circuit if amount is 0.
+        if (amount == 0) return;
+
+        let staker_address = signer::address_of(staker);
+        assert_staking_contract_exists(staker_address, operator);
+
+        let store = borrow_global_mut<Store>(staker_address);
+        let staking_contract = simple_map::borrow_mut(&mut store.staking_contracts, &operator);
+
+        // Force distribution of any already inactive stake.
+        distribute_internal(staker_address, operator, staking_contract, &mut store.distribute_events);
+
+        // For simplicity, we request commission to be paid out first. This avoids having to ensure to staker doesn't
+        // withdraw into the commission portion.
+        let commission_paid = request_commission_internal(
+            operator,
+            staking_contract,
+            &mut store.add_distribution_events,
+            &mut store.request_commission_events,
+        );
+
+        // If there's less active stake remaining than the amount requested (potentially due to commission),
+        // only withdraw up to the active amount.
+        let (active, _, _, _) = stake::get_stake(staking_contract.pool_address);
+        if (active < amount) {
```

Comment:
would this cause any confusion when I request to unlock 100 coin, request executed successfully, but I only got 99 as a result?

Reply:
Yes unfortunately it might be a bit confusing due to the commission. Technically without doing the math before calling this, the staker can't know whether they'd get less as commission will be charged first, which can leave a remaining amount of less than the amount the staker is requesting.